flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandlerTest.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io;
+
+import org.apache.flink.runtime.checkpoint.CheckpointMetaData;
+import org.apache.flink.runtime.checkpoint.CheckpointMetrics;
+import org.apache.flink.runtime.checkpoint.CheckpointOptions;
+import org.apache.flink.runtime.checkpoint.CheckpointType;
+import org.apache.flink.runtime.checkpoint.channel.ChannelStateWriter;
+import org.apache.flink.runtime.event.TaskEvent;
+import org.apache.flink.runtime.io.network.api.CheckpointBarrier;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.consumer.InputChannel;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate;
+import org.apache.flink.runtime.io.network.partition.consumer.SingleInputGateBuilder;
+import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
+import org.apache.flink.runtime.operators.testutils.DummyEnvironment;
+import org.apache.flink.runtime.state.CheckpointStorageLocationReference;
+import org.apache.flink.util.function.ThrowingRunnable;
+
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static org.apache.flink.runtime.checkpoint.CheckpointType.CHECKPOINT;
+import static org.apache.flink.runtime.checkpoint.CheckpointType.SAVEPOINT;
+import static org.apache.flink.runtime.io.network.api.serialization.EventSerializer.toBuffer;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+/**
+ * {@link AlternatingCheckpointBarrierHandler} test.
+ */
+public class AlternatingCheckpointBarrierHandlerTest {
```

Comment:
Good tests! But I think you also need to test `BufferReceivedListener` as the current version of AlternatingCheckpointBarrierHandler is probably not allowing any UC on input side.

Reply:
Thanks! I'll test that `InputChannel`s don't notify `BufferReceivedListener`.

Reply:
I was more concerned that the alternation is currently not working when a barrier is received through notify. In particular, an UC is only triggered when the first barrier is processed vs. received defeating the purpose of UC.

Reply:
`AlternatingCheckpointBarrierHandler` will receive the barrier as usual (via "`pollNext`") and switch its aligner. I don't see any reason to notify it immediately (or through the mailbox) as soon as the channel receives a barrier.

Do you see any problems with this approach?

Reply:
There is an inconsistency issue from perspective of InputChannels threads. But they are accessing `AlternatingCheckpointBarrierHandler` only via the `notifyBarrierReceived` and `notifyBufferReceived` and those methods are only called/active AFTER unaligned checkpoint has started in the task thread, so barrier had to be already processed by the `AlternatingCheckpointBarrierHandler`. Hence `activeHandler` had to already switch `unalignedHandler`.

This is very fragile contract, but :

- solution would require to synchronize code even more 
- I think there is no bug at the moment
- once we fix the threading model and get rid of the listener all together, the problem will be gone

I only hope I'm not missing something.

Reply:
> There is an inconsistency issue from perspective of InputChannels threads. But they are accessing AlternatingCheckpointBarrierHandler only via the notifyBarrierReceived and notifyBufferReceived

If I got your point right,
`InputChannels` threads notify `CheckpointBarrierUnaligner.ThreadSafeUnaligner`, not `AlternatingCheckpointBarrierHandler`.
And `ThreadSafeUnaligner` doesn't know about the `AlternatingCheckpointBarrierHandler`.

There is "inconsistency" in the sense that `AlternatingCheckpointBarrierHandler` can "lag behind" `ThreadSafeUnaligner` but it shouldn't cause any problems.

Reply:
> There is "inconsistency" in the sense that AlternatingCheckpointBarrierHandler can "lag behind" ThreadSafeUnaligner but it shouldn't cause any problems.

Yes, me and Arvid think and hope that's the case. (I've synced with him offline)